h5py/_hl/base.py

```diff
@@ -75,10 +81,7 @@ def guess_dtype(data):
         if isinstance(data, h5r.Reference):
             return h5t.ref_dtype
 
-        item_type = object_collection_of(data)
-        if item_type is None:
-            # Potentially scalar
-            item_type = type(data)
+        item_type = find_item_type(data)
```

Comment:
There is a path through `find_item_type` that returns `None`, but the code that used to be here ensured that `item_type` was never `None` before going into the rest of the code.

Reply:
It did, but then the only things we do with item_type are to check `item_type is bytes` and `item_type is str` immediately below. The case which the extra `if item_type is None` previously handled - a single str/bytes object - is now handled inside `find_item_type()`.

`item_type` will be None if there is a non-homogeneous collection (different types in a list) or a numpy array with a more specific dtype already.